File: backend/coffeeing/src/main/java/com/ssafy/coffeeing/modules/product/domain/Capsule.java
package com.ssafy.coffeeing.modules.product.domain;

import com.ssafy.coffeeing.modules.global.embedded.CoffeeCriteria;
import com.ssafy.coffeeing.modules.util.base.BaseEntity;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

import javax.persistence.AttributeOverride;
import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Entity;

@Getter
@SuperBuilder
@AllArgsConstructor(access = AccessLevel.PROTECTED)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AttributeOverride(name = "id", column = @Column(name = "capsule_id"))
@Entity
public class Capsule extends BaseEntity {

    @Column
    private String brandKr;

    @Column
    private String brandEng;

    @Column
    private String capsuleNameKr;

    @Column
    private String capsuleNameEng;

    @Embedded
    private CoffeeCriteria coffeeCriteria;

    @Column
    private String flavorNote;

    @Column
    private Integer machineType;

    @Column
    private String imageUrl;

    @Column
    private String productDescription;

    @Column
    @Builder.Default
    private Integer totalScore = 0;

    @Column
    @Builder.Default
    private Integer totalReviewer = 0;

    @Column
    @Builder.Default
    private Integer popularity = 0;

    public void addReview(Integer score) {
        this.totalScore += score;
        this.totalReviewer++;

        this.popularity += score;
    }

    public void deleteReview(Integer score) {
        this.totalScore -= score;
        this.totalReviewer--;

        this.popularity += score;
    }

    public void updateReview(Integer diff) {
        this.totalScore += diff;

        this.popularity += diff;
    }

    public void attachFeed() {
        this.popularity += 10;
    }

    public void detachFeed() {
        this.popularity -= 10;
    }
}

File: model/coffeeing/app/database/models/capsule.py
from sqlalchemy import Column, ForeignKey
from sqlalchemy import BIGINT, Integer, String, DateTime, Double, Text
from sqlalchemy.orm import relationship

from ..connection import Base

class Capsule(Base):
    def __getitem__(self, key):
        return getattr(self, key)

    def __setitem__(self, key, value):
        return setattr(self, key, value)

    def keys(self):
        return list(self.__dict__.keys())[1:]

    def values(self):
        return list(self.__dict__.values())[1:]

    __tablename__ = "capsule"

    capsule_id = Column(BIGINT, primary_key=True, index=True)
    created_at = Column(DateTime())
    modified_at = Column(DateTime())
    brand_eng = Column(String(255))
    brand_kr = Column(String(255))
    capsule_name_eng = Column(String(255))
    capsule_name_kr = Column(String(255))
    acidity = Column(Double)
    body = Column(Double)
    roast = Column(Double)
    flavor_note = Column(String(255))
    machine_type = Column(Integer)
    image_url = Column(Text)
    product_description = Column(Text)
    total_reviewer = Column(Integer)
    total_score = Column(Integer)
    popularity = Column(Integer)